### PR TITLE
feat(logs): generate curl command from log details

### PR DESCRIPTION
**Pull request by RTabary**
You can check the result.
I have a last and blocking issue: the api-key header is not set on the ClientRequest.headers, so i can get the value from the log object but i can't guess the name of the header.

### CODE REVIEW

**Comment by brasseld** [comment]
Hi @romsalto 

Really nice job!

However, it seems that some parts are missing to generate a proper curl command :)

Capture d’écran 2019-03-16 à 09 53 17

**Comment by brasseld** [comment]
Also, I need to see why do we have the X-Gravitee-Transaction-Id from the consumer request... seems it should be set by the gateway, this header must not be part of the incoming request....

**Comment by brasseld** [comment]
Still me :)

When I click on 'Hide curl', it takes some time (let's say 1 second) to effectively hide curl command. Do you know why ?

File: src/management/api/analytics/logs/log.component.ts [hunk 1]
```diff
@@ -36,6 +38,42 @@ const LogComponent: ng.IComponentOptions = {
 
       return null;
     };
+    this.getCurl = function(log, scope){
+      var request = (scope == "client") ? log.clientRequest : log.proxyRequest;
+      if(scope == "client"){
+        var uri = request.headers['X-Forwarded-Proto']+"://"+request.headers['Host']+request.headers['X-Forwarded-Prefix']+request.uri;
```
Comment (brasseld) [inline src/management/api/analytics/logs/log.component.ts]: `request.headers['X-Forwarded-Prefix']` is not always provided, you should take care about it and check if the header is existing or not.

Comment (brasseld) [reply src/management/api/analytics/logs/log.component.ts]: The same rule must be applied for `request.headers['X-Forwarded-Proto']`

Comment (romsalto) [reply src/management/api/analytics/logs/log.component.ts]: Hi @brasseld you're right i forgot that point, but i'm not sur of how to handle that...
```suggestion
        var proto = (request.headers['X-Forwarded-Proto']) ? request.headers['X-Forwarded-Proto']+"://" : "";
        var prefix = (request.headers['X-Forwarded-Prefix']) ? request.headers['X-Forwarded-Prefix'] : "";
        var uri =proto+request.headers['Host']+prefix+request.uri;
```
Would it be ok ?

Comment (brasseld) [reply src/management/api/analytics/logs/log.component.ts]: I'm not sure about `proto` part. IMHO, there is no way currently to get the scheme of the request if you're not relying on the `X-Forwarded-Proto`. Meaning that there is probably something to do in Gravitee to take care of the proto and save it as part of the request's log.

The other point is about the `Host` header. If you're relying on `X-Forwarded-Proto`, I suppose you would also have to extract the value of `X-Forwarded-Host` if this later is provided, no?

Ok, for the path-prefix.

Comment (romsalto) [reply src/management/api/analytics/logs/log.component.ts]: I'm totally agree with you...
The best solution is to have all of these informations without reading the headers.
I would have loved helping for the gravitee back part, but I don't feel comfortable with Java (yet)...

Comment (brasseld) [reply src/management/api/analytics/logs/log.component.ts]: No worries, I'm understanding quite perfectly your position... and I know I have to decline all Java code you may provide in the future :-P

Let me see what can we do to manage proto and host.

**Comment by romsalto** [comment]
> When I click on 'Hide curl', it takes some time (let's say 1 second) to effectively hide curl command. Do you know why ?

Yes i saw that too, but i have no idea why it is so long to hide.
I tried with other solutions before submitting the PR like using ng-class to handle the visibility instead of ng-if, the result is the same. this is weird because the "show" is fast but the "hide" is long while both uses the same logic.
i keep looking for a optimization.

> I have a last and blocking issue: the api-key header is not set on the ClientRequest.headers, so i can get the value from the log object but i can't guess the name of the header.

On that point, is it possible to have the api-key header in the request.headers ? Or any way to get the header name ?

**Comment by gaetanmaisse** [comment]
@RTabary I'm closing this PR as it is 2 years old and has a lot of conflicts, feels free to reopen it and comment if you think it is still valuable 🙂 .

**Pull request closed without merging**